driver-core/src/main/com/mongodb/internal/connection/BsonWriterHelper.java

```diff
@@ -70,12 +71,14 @@ static void writePayload(final BsonWriter writer, final BsonOutput bsonOutput, f
     }
 
     private static boolean writeDocument(final BsonWriter writer, final BsonOutput bsonOutput, final MessageSettings settings,
-                                         final BsonDocument document, final int messageStartPosition, final int batchItemCount) {
+                                         final BsonDocument document, final int messageStartPosition, final int batchItemCount,
+                                         final int maxSplittableDocumentSize) {
         int currentPosition = bsonOutput.getPosition();
         getCodec(document).encode(writer, document, ENCODER_CONTEXT);
         int messageSize = bsonOutput.getPosition() - messageStartPosition;
         int documentSize = bsonOutput.getPosition() - currentPosition;
-        if (exceedsLimits(settings, messageSize, documentSize, batchItemCount)) {
+        if (exceedsLimits(settings, messageSize, documentSize, batchItemCount)
+                || (batchItemCount > 1 && bsonOutput.getPosition() - messageStartPosition > maxSplittableDocumentSize)) {
```

Comment:
This is the main change.  The rest is just passing the necessary limits to this spot.